client/src/main/java/io/pravega/client/tables/KeyValueTable.java

```diff
@@ -98,197 +102,147 @@
+     * <p>
+     * If {@link TableEntry#isVersioned()} returns true, this will be a:
+     * <ul>
+     * <li> Conditional Insert if the {@code entry}'s {@link TableEntry#getKey()} {@link TableKey#getVersion()} equals
+     * {@link Version#NOT_EXISTS} (i.e., if the {@link TableEntry} was created using {@link TableEntry#notExists}).
+     * The {@link TableEntry} will only be inserted if there doesn't exist a {@link TableKey} in the {@link KeyValueTable}
+     * matching the given {@code entry}'s {@link TableEntry#getKey()}).
+     * <li> Conditional Update if the {@code entry}'s {@link TableEntry#getKey()} {@link TableKey#getVersion()} does not
+     * equals {@link Version#NOT_EXISTS} or {@link Version#NO_VERSION} (i.e., if the {@link TableEntry} was created using
+     * {@link TableEntry#versioned}). The {@link TableEntry} will only be inserted if there exists a {@link TableKey} in
+     * the {@link KeyValueTable} matching the given {@code entry}'s {@link TableEntry#getKey()} and both key versions match.
+     * </ul>
+     * <p>
+     * All other invocations will be executed as Unconditional Updates or Insertions ({@code entry.getKey().getVersion()} equals
+     * {@link Version#NO_VERSION} - such as when the entry was created using {@link TableEntry#unversioned}).
      *
-     * @param keyFamily (Optional) The Key Family for the Entry. If null, this Entry will not be associated with any
-     *                  Key Family.
-     * @param key       The Key to insert or update.
-     * @param value     The Value to be associated with the Key.
+     * @param entry The {@link TableEntry} to insert or update.
      * @return A CompletableFuture that, when completed, will contain the {@link Version} associated with the newly
      * inserted or updated entry.
      */
-    CompletableFuture<Version> put(@Nullable String keyFamily, @NonNull KeyT key, @NonNull ValueT value);
-
-    /**
-     * Conditionally inserts a new Entry in the {@link KeyValueTable} if the given Key is not already present.
-     *
-     * @param keyFamily (Optional) The Key Family for the Entry. If null, this Entry will not be associated with any
-     *                  Key Family.
-     * @param key       The Key to insert.
-     * @param value     The Value to be associated with the Key.
-     * @return A CompletableFuture that, when completed, will contain the {@link Version} associated with the newly
-     * inserted or updated entry. Notable exceptions:
-     * <ul>
-     * <li>{@link ConditionalTableUpdateException} If the Key is already present in the {@link KeyValueTable} for the
-     * provided Key Family.
-     * See the {@link KeyValueTable} doc for more details on Conditional Update Responses.
-     * </ul>
-     */
-    CompletableFuture<Version> putIfAbsent(@Nullable String keyFamily, @NonNull KeyT key, @NonNull ValueT value);
+    CompletableFuture<Version> put(@NonNull TableEntry entry);
```

Comment:
Previously the structure and verbage of these APIs was explicitly modeled on those of [ConcurrentMap](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/concurrent/ConcurrentMap.html)

Is there some reason that you want to overload the `put` method instead?

Reply:
From user's feedback, that was a rather complicated API that caused more confusion than it helped. 

So a few things were done to this API to simplify it:
- Remove the parametrized types and exposing `ByteBuffer` for keys and values.
- Coalesced redundant APIs (like the ones you mentioned - inspired from Java's Map).

This reduces the API to a bare-bones one which allows for greatest flexibility (from a user's POV). In a subsequent PR/release (not yet planned), it should be trivial to add a thin layer on top of this API for those users that do want a more familiar API and/or parametrized types. Translating from a Map-style API to this one is trivial and can be done at a later time without needing any changes underneath.

Reply:
I think this is covered by the commit I pushed on April 30.